Userland/Libraries/LibWeb/Fetch/Response.cpp

```diff
@@ -0,0 +1,309 @@
+/*
+ * Copyright (c) 2022, Linus Groh <linusg@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/URLParser.h>
+#include <LibWeb/Bindings/MainThreadVM.h>
+#include <LibWeb/Fetch/Enums.h>
+#include <LibWeb/Fetch/Infrastructure/HTTP/Bodies.h>
+#include <LibWeb/Fetch/Infrastructure/HTTP/Responses.h>
+#include <LibWeb/Fetch/Infrastructure/HTTP/Statuses.h>
+#include <LibWeb/Fetch/Response.h>
+#include <LibWeb/HTML/Scripting/Environments.h>
+#include <LibWeb/HTML/Window.h>
+#include <LibWeb/Infra/JSON.h>
+
+namespace Web::Fetch {
+
+Response::Response(JS::Realm& realm, NonnullOwnPtr<Infrastructure::Response> response)
+    : PlatformObject(realm)
+    , m_response(move(response))
+{
+    auto& window = verify_cast<HTML::Window>(realm.global_object());
+    set_prototype(&window.cached_web_prototype("Response"));
+}
+
+Response::~Response() = default;
+
+void Response::visit_edges(Cell::Visitor& visitor)
+{
+    Base::visit_edges(visitor);
+    visitor.visit(m_headers);
+}
+
+// https://fetch.spec.whatwg.org/#concept-body-mime-type
+// https://fetch.spec.whatwg.org/#ref-for-concept-header-extract-mime-type%E2%91%A7
+Optional<MimeSniff::MimeType> Response::mime_type_impl() const
+{
+    // Objects including the Body interface mixin need to define an associated MIME type algorithm which takes no arguments and returns failure or a MIME type.
+    // A Response object’s MIME type is to return the result of extracting a MIME type from its response’s header list.
+    return m_response->header_list()->extract_mime_type();
+}
+
+// https://fetch.spec.whatwg.org/#concept-body-body
+// https://fetch.spec.whatwg.org/#ref-for-concept-body-body%E2%91%A8
+Optional<Infrastructure::Body const&> Response::body_impl() const
+{
+    // Objects including the Body interface mixin have an associated body (null or a body).
+    // A Response object’s body is its response’s body.
+    return m_response->body().has_value()
+        ? m_response->body().value()
+        : Optional<Infrastructure::Body const&> {};
+}
+
+// https://fetch.spec.whatwg.org/#concept-body-body
+// https://fetch.spec.whatwg.org/#ref-for-concept-body-body%E2%91%A8
+Optional<Infrastructure::Body&> Response::body_impl()
+{
+    // Objects including the Body interface mixin have an associated body (null or a body).
+    // A Response object’s body is its response’s body.
+    return m_response->body().has_value()
+        ? m_response->body().value()
+        : Optional<Infrastructure::Body&> {};
+}
+
+// https://fetch.spec.whatwg.org/#reponse-create
+JS::NonnullGCPtr<Response> Response::create(NonnullOwnPtr<Infrastructure::Response> response, Headers::Guard guard, JS::Realm& realm)
+{
+    auto& window = verify_cast<HTML::Window>(realm.global_object());
+
+    // Copy response's header list before it's being move()'d.
+    auto response_reader_list = response->header_list();
```

Comment:
One of those header list copy before moves

Reply:
This is just a RefPtr copy, will clarify the comment.